gazebo_plugins/test/test_gazebo_ros_wheelslip.cpp

```diff
@@ -0,0 +1,276 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gazebo/test/ServerFixture.hh>
+#include <rclcpp/rclcpp.hpp>
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+using namespace std::chrono_literals;
+
+/// Tests the gazebo_ros_wheelslip plugin
+class GazeboRosWheelSlipTest : public gazebo::ServerFixture
+{
+public:
+  void test_parameters(
+    std::string world_file,
+    std::string node_name,
+    std::map<std::string, double> parameter_pairs)
+  {
+    // Load the world
+    this->Load(world_file, true);
+    auto world = gazebo::physics::get_world();
+    ASSERT_NE(nullptr, world);
+    auto node = std::make_shared<rclcpp::Node>("test_gazebo_ros_wheelslip");
+    ASSERT_NE(nullptr, node);
+
+    // Create parameters client
+    auto parameters_client = std::make_shared<rclcpp::SyncParametersClient>(
+      node, node_name);
+
+    // Wait for serive to show up
+    int counter = 0;
+    while (!parameters_client->wait_for_service(1s)) {
+      std::cout << "service not available, waiting again..." << std::endl;
+      counter++;
+      if (counter > 5) {FAIL();}
+    }
+
+    // Verify the parameters
+    std::vector<std::string> parameter_names;
+    for (auto & element : parameter_pairs) {
+      parameter_names.push_back(element.first);
+    }
+
+    /* Query for the parameters and check if they match with the expected ones.
+       Plugin takes a while to set the parameters, so we make multiple attempts to
+       check the parameters */
+    int max_attempts = 5;
+    bool flag_do_parameters_match;
+    for (int i = 0; i < max_attempts; i++) {
+      // Assume all expected and received parameters match
+      flag_do_parameters_match = true;
+
+      // Query for parameters
+      auto parameters_received = parameters_client->get_parameters(parameter_names);
+      for (auto & parameter : parameters_received) {
+        if (parameter_pairs[parameter.get_name()] != parameter.as_double()) {
+          // Parameters don't match
+          flag_do_parameters_match = false;
+          break;
+        }
+      }
+
+      // Check the flag_do_parameters_match, if it was true, no need to query again
+      if (flag_do_parameters_match) {break;}
+      std::this_thread::sleep_for(0.2s);
+    }
```

Comment:
Here's a solution where we avoid sleeping altogether (I haven't actually tested if this compiles :grimacing:):

```c++
    bool all_parameters_match = false;
    auto start_time = std::chrono::steady_clock::now();
    while (!all_parameters_match && (std::chrono::steady_clock::now() - start_time) < 1s) {
      all_parameters_match = true;

      // Query for parameters
      auto parameters_received = parameters_client->get_parameters(parameter_names);
      for (auto & parameter : parameters_received) {
        if (parameter_pairs[parameter.get_name()] != parameter.as_double()) {
          all_parameters_match = false;
          break;
        }
      }
    }
```

If the time we need to wait if very short, then I _think_ this sleepless wait is more efficient. I don't feel strongly one way or the other, just suggesting an alternative.